src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -221,40 +208,36 @@ static TimestampWrapper validateAndParseTimestampTz(Object input, Map<String, St
    *     scale
    */
   static TimestampWrapper validateAndParseTimestampTz(Object input, int scale) {
-    try {
-      if (input instanceof String) {
-        String stringInput = (String) input;
-        SFTimestamp timestamp = snowflakeDateTimeFormatter.parse(stringInput);
-        if (timestamp == null) {
-          throw new SFException(
-              ErrorCode.INVALID_ROW, input, "Cannot be converted to a timestamp_tz value");
-        }
-
-        int fraction = getFractionFromTimestamp(timestamp);
 
-        BigDecimal epochTimeAtSecondsScale =
-            timestamp.getNanosSinceEpoch().divide(BigDecimal.valueOf(Power10.intTable[9]));
-        if (fraction % Power10.intTable[9 - scale] != 0) {
-          throw new SFException(
-              ErrorCode.INVALID_ROW,
-              input.toString(),
-              String.format(
-                  "Row specifies accuracy greater than column scale.  fraction=%d, scale=%d",
-                  fraction, scale));
-        }
-        TimestampWrapper wrapper =
-            new TimestampWrapper(
-                timestamp.getSeconds().longValue(),
-                fraction,
-                getTimeInScale(getStringValue(epochTimeAtSecondsScale), scale),
-                timestamp);
-        return wrapper;
-      } else {
-        throw new SFException(
-            ErrorCode.INVALID_ROW, input, "Cannot be converted to a timestamp_tz value");
-      }
-    } catch (NumberFormatException e) {
-      throw new SFException(ErrorCode.INVALID_ROW, input.toString(), e.getMessage());
+    String stringInput;
+    if (input instanceof String) stringInput = (String) input;
+    else if (input instanceof LocalDate) stringInput = input.toString();
+    else if (input instanceof LocalDateTime) stringInput = input.toString();
+    else if (input instanceof ZonedDateTime)
+      stringInput =
+          DateTimeFormatter.ISO_DATE_TIME.format(((ZonedDateTime) input).toOffsetDateTime());
+    else if (input instanceof OffsetDateTime)
+      stringInput = DateTimeFormatter.ISO_DATE_TIME.format((OffsetDateTime) input);
+    else
+      throw new SFException(
+          ErrorCode.INVALID_ROW, input.toString(), "Data type not supported for timestamp");
```

Comment:
looks like this logic could be moved to a helper to avoid duplication

Reply:
Done